src/main/java/atmosfears/AtmosFearsBE/service/Predictor.java

```diff
@@ -0,0 +1,73 @@
+package atmosfears.AtmosFearsBE.service;
+
+import com.github.rcaller.rstuff.RCaller;
+import com.github.rcaller.rstuff.RCode;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import java.util.stream.DoubleStream;
+
+@Service
+public class Predictor {
+
+    public List<Double> predict(String csv_file_path){
+
+        // TODO odpalanie wspólnego rCallera dla wszystkich wywołań predict'u
+
+        // MODEL WILL NOT WORK IF THE NECESSARY R PACKAGES ARE NOT INSTALLED ALONGSIDE R ITSELF
+        // For detail about how to install the packages refer to README
+
+        // EXAMPLE OF ARGUMENTS IN USE:
+        // double[] pred = Predictor.predict("src/main/data/example_sensor_data.csv");
+
+        try{
+
+            // R ENVIRONMENT INITIALIZATION
+            RCaller rCaller = RCaller.create();
+            RCode code = RCode.create();
+
+            // LOADING OF THE LIBRARIES
+            code.addRCode("library(tidyverse)");
+            code.addRCode("library(tidymodels)");
+            code.addRCode("library(openair)");
+            code.addRCode("tidymodels_prefer()");
+            code.addRCode("library(\"Cubist\")");
+            code.addRCode("library(\"reshape2\")");
+            code.addRCode("library(\"plyr\")");
+            code.addRCode("library(\"rules\")");
+            code.addRCode("library(\"rio\")");
+            code.addRCode("library(\"openxlsx\")");
+            code.addRCode("library(\"zip\")");
+            code.addRCode("conflicted::conflicts_prefer(plyr::mutate)");
+
+            // LOADING OF MODEL
+            code.addRCode("load(\"src/main/data/model.Rdata\")");
+
+            // LOADING OF DATA
+            code.addRCode("ops_data <- rio::import(file = \"" + csv_file_path + "\", setclass = \"tbl_df\")");
+            code.addRCode("ops_data <- ops_data |> mutate(wday = as.factor(wday), hour = as.factor(hour))");
+
+            // PERFORMING "CALIBRATION"
+            code.addRCode("out <- bind_cols(ops_data, predict(model$.workflow[[1]], ops_data, type = \"numeric\"))");
+            code.addRCode("out <- out |> select(.pred)");
+
+            // RETRIEVING RESULT
+            rCaller.setRCode(code);
```

Comment:
I think it will be better to store R code in a separate file and read it from a file. This way od adding code is very repetitive and a bit unreadable.

I would suggest the following steps:
1. Create a `scripts` directory inside the `main` folder (at the same level of nesting where `data` directory leaves).
2. Create a file with the R script,
3. Create a `utils` directory inside the `main/java` folder and add a file reader class. It can look similar to this:
```java
package atmosfears.AtmosFearsBE.utils;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

public class FileReader {
    public static String readFile(String path)
            throws FileNotFoundException {
        File file = new File(path);
        Scanner scanner = new Scanner(file);
        String content = scanner.useDelimiter("\\Z").next();
        scanner.close();
        return content;
    }
}
```
4. Simplify the `Predictor` class code. For example, it can be:
```java
@Service
public class Predictor {
    private static final String R_SCRIPT_PATH = "src/main/scripts/code.r";
    private static final String R_SCRIPT_CODE;

    private static final RCaller rCaller = RCaller.create();

    static {
        try {
            R_SCRIPT_CODE = FileReader.readFile(R_SCRIPT_PATH);
        } catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        }
    }

    public List<Double> predict(String csv_file_path) {
        // TODO odpalanie wspólnego rCallera dla wszystkich wywołań predict'u

        // MODEL WILL NOT WORK IF THE NECESSARY R PACKAGES ARE NOT INSTALLED ALONGSIDE R ITSELF
        // For detail about how to install the packages refer to README

        // EXAMPLE OF ARGUMENTS IN USE:
        // double[] pred = Predictor.predict("src/main/data/example_sensor_data.csv");
        try {
            // LOAD R SCRIPT
            String rCode = R_SCRIPT_CODE.replace("{%CSV_DATA_FILE%}",  csv_file_path);

            // R ENVIRONMENT INITIALIZATION
            RCode code = RCode.create();
            code.addRCode(rCode);

            // RETRIEVING RESULT
            rCaller.setRCode(code);
            rCaller.runAndReturnResult("out");

            return DoubleStream.of(rCaller.getParser()
                                          .getAsDoubleArray("_pred"))
                               .boxed()
                               .collect(Collectors.toCollection(ArrayList::new));
        } catch (Exception e) {
            Logger.getLogger(examples.Main.class.getName())
                  .log(Level.SEVERE, e.getMessage());
        }

        return null;
    }
}
```
The main change in the `Predictor` class is that you load the R script code from a file only once and create the static `rCaller` variable that will be reused every time when the `predict` method is called.

You shouldn't load the whole script and create caller on each `predict` call as it will be very inefficient.

To use custom CSV file path in the R script, you can add some placeholder, such as:
```r
ops_data <- rio::import(file = "{%CSV_DATA_FILE%}", setclass = "tbl_df")
```
in the R code and then use `.replace()` string method in the Java code to use a proper file path:
```java
String rCode = R_SCRIPT_CODE.replace("{%CSV_DATA_FILE%}",  csv_file_path);
```

I think it should work properly and be much cleaner.

Reply:
Thank you for the very precise suggestions - I will implement them soon but I don't think I'll able to commit them today.
If you can, please approve this PR before our meeting with client today :D